text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,879 @@
+- `maximum`: behavior today (default for v1 and v2 resolvers)
+  - Needed for [verifying latest dependencies](https://doc.rust-lang.org/nightly/cargo/guide/continuous-integration.html#verifying-latest-dependencies)
+- `minimum` (unstable): `-Zminimal-versions`
+  - As this is just precedence, `-Zdirect-minimal-versions` doesn't fit into this
+- `rust-version`:  what is defined in the package (default for v3 resolver)
+- `rust-version=` (future possibility)
+  - `package`: long form of `rust-version`
+  - `rustc`: the current running version
+    - Needed for "separate development / publish MSRV" workflow
+  - `<x>[.<y>[.<z>]]`: manually override the version used
+
+If a `rust-version` value is used, we'd switch to `maximum` when `--ignore-rust-version` is set.
+This will let users effectively pass `--ignore-rust-version` to all commands,
+without having to support the flag on every single command.
+
+## `cargo build`
+
+The MSRV-compatibility build check will be demoted from an error to a `deny`-by-default workspace
+[diagnostic](https://github.com/rust-lang/cargo/issues/12235),
+allowing users to intentionally use dependencies on an unsupported (or less supported) version of Rust
+without requiring `--ignore-rust-version` on every invocation.
+
+## `cargo update`
+
+`cargo update` will inform users when an MSRV or semver incompatible version is available.
+`cargo update -n` will also report this information so that users can check on the status of this at any time.
+
+**Note:** other operations that cause `Cargo.lock` entries to be changed (like
+editing `Cargo.toml` and running `cargo check`) may not inform the user.
+If they want to check the status of things, they can run `cargo update -n`.
+
+Users may pass
+- `--ignore-rust-version` to pick the latest dependencies, ignoring all `rust-version` fields (your own and from dependencies)
+- `--update-rust-version` to pick the `rustc --version`-compatible dependencies, updating your `package.rust-version` if needed to match the highest of your dependencies
+
+Those flags will also be added to `cargo generate-lockfile`
+
+## `cargo add`
+
+`cargo add <pkg>` (no version) will pick a version requirement that is low
+enough so that when it resolves, it can pick a dependency that is
+MSRV-compatible.
+`cargo add` will warn when it does this.
+
+Users may pass
+- `--ignore-rust-version` to pick the latest dependencies, ignoring all `rust-version` fields (your own and from dependencies)
+- `--update-rust-version` to pick the `rustc --version`-compatible dependencies, updating your `package.rust-version` if needed to match the highest of your dependencies
+
+## `cargo publish`
+
+We will add `"auto"` for `package.rust-version`.
+On `cargo publish` / `cargo package`, the generated `*.crate`s `Cargo.toml` will have `"auto"` replaced with `rustc --version`.
+If `rustc --version` is a pre-release, it will be left as unspecified.
+
+`cargo new` will include `package.rust-version = "auto"`.
```

Comment:
@seanmonstar 
> > We will add "auto" for `package.rust-version`. `cargo new` will include `package.rust-version = "auto"`.
> 
> I couldn't find an explanation for _why_. Is this really needed? It doesn't seem like it should be. It feels like with it being defaulted on, new crates would just have an MSRV chosen for them. While it would certainly be true that the crate could be compiled with that version, it shouldn't mean it couldn't be compiled with an older one. Or that it should "hold back" any resolution to a newer one. That seems like something that should be explicitly chosen by the author.
> 
> Perhaps I misunderstand. Regardless, if it _is_ needed, that should be in the RFC text.

Reply:
Most of the rationale gets summarized in the [comparison between resolver solutions](https://github.com/epage/rfcs/blob/msrv/text/3537-msrv-resolver.md#resolver-behavior) (as a lot of this is intertwined)

To expand on the one-line entry: An MSRV-aware resolver is of limited use if there is not MSRV data in the Index.  Originally I punted on this with different potential ideas for inferring this.  When stepping back and looking at the [principles](https://github.com/epage/rfcs/blob/msrv/text/3537-msrv-resolver.md#workflows) for evaluating the workflows, a key part was what we consider "support".  Previously, Cargo hasn't had much to say on this and it has meant a lot of things to a lot of people.  In talking over this in the team and in working through the principles, its solidified that "support" is an explicit choice by a maintainer and not something that can be inferred syntactically. 

Yes, this means that MSRVs will err on the side of being "too high".  Again, this is about what the maintainer is deciding they are willing to support.  However, we are providing an escape hatch.  The RFC also calls for downgrading the "incompatible MSRV" error into a deny-diagnostic allowing dependents can explicitly opt-in to using packages on unsupported toolchain versions.

Reply:
So it seems like the lines can be summarized as:

As a principle it is better that users by default gets an explicit error indicating that they need to upgrade to Rust N-0*, rather than naively building a crate. The rationale being that naively building a crate might result in less explicit build failures.

Is this accurate?

> *: I characterize it as "should upgrade to N - 0", because it seems like the proposed scheme will cause the error to be caused by any crate in a tree of dependencies which does for whatever reason not specify a `rust-version`. The probability of it happening for any given build then seems fairly high unless dependencies are consistently being held back.

Reply:
I read through all of it, and still wasn't sure why this particular part of the RFC was needed. The intertwining was not clear to me. It seems like the title feature can be provided without auto (otherwise all older crates won't work).

I don't think automatically picking an MSRV is the right move. It also seems like it doesn't _depend_ on the resolver part, so it could be a separate RFC.

Reply:
> As a principle it is better that users by default gets an explicit error indicating that they need to upgrade to Rust N-0*, rather than naively building a crate. The rationale being that naively building a crate might result in less explicit build failures.

Not quite.  More 

As a principle it is better that users by default gets an older dependency, with a warning that they need to upgrade to Rust N-0*, rather than naively building a crate on a Rust version the maintainer of the crate does not support and may give less explicit build failures.  The user may explicitly opt-in to unsupported versions by changing the error into a warning or hiding it completely via `[lints.cargo]`.

Reply:
> I read through all of it, and still wasn't sure why this particular part of the RFC was needed. The intertwining was not clear to me. It seems like the title feature can be provided without auto (otherwise all older crates won't work).
>
> I don't think automatically picking an MSRV is the right move. It also seems like it doesn't depend on the resolver part, so it could be a separate RFC.

I'm normally a big advocate for small RFCs.  However, when looking over the problem and the solutions and how they reinforce each other, combined with where the discussions have gone, I felt it was important that we at least look at the solution in aggregate so we have a feel for what an incremental step will be.

Yes, this RFC doesn't do anything to backfill crates with no MSRV set.  I don't think that invalidates that making sure the information is there going forward will help in the medium and long term.  It likely will even help in the short term as users won't be on this resolver until 2024 edition, by default, and I propose we stabilize these independently, making it so we could quickly get `auto` in.

Reply:
Thanks.

> rather than naively building a crate on a Rust version the maintainer of the crate does not support

Why did you make this language explicit? As a general principle I would expect that if a crate compiles it is "supported" regardless of intent. Do you see any downstream issues with implicitly supported crates apart from possibly obscure compile errors?

Reply:
> IIUC It must be an error if the old resolver strategy (maximum) is used, and this behavior if the new one (rust-version) is used right?

If using maximal and running on an old toolchain, yes.

> Regardless I think a reasonable conclusion is that it's creating a scenario where ecosystem pressure to upgrade to N-0 is applied. 

Could you expand on why you feel this is putting pressure on everyone to be N-0?  It is moving packages with a previously undefined policy to N-0 but that undefined policy was unpredictable and was roughly the same result imo.  And you already had to deal packages having newer MSRVs, whether by not using them or managing a lockfile.   For me, I see this as making it less onerous to maintain a N-M policy than it was before

> So is this pressure good or bad? Is it good or bad for different user personas?

imo we need some level of pressure to move forward to avoid stagnation but it has to be subtle to not alienate people who do need to work with old rust versions.

> Why did you make this language explicit? As a general principle I would expect that if a crate compiles it is "supported" regardless of intent. Do you see any downstream issues with implicitly supported crates apart from possibly obscure compile errors if the crate effectively bumps its rust-version?

When discussing [workflows](https://github.com/epage/rfcs/blob/msrv/text/3537-msrv-resolver.md#workflows), we derive from them that "support" is a matter of maintainer intention and effort.  For example, the maintainer should be doing compile checks and maybe even running tests on the oldest version they support.  If they are not intending to support that prior versions then they aren't putting in this effort and it exists in an unknown state.  To ensure the quality of the experience of users, we want the default case to avoid these unknown states.

Reply:
> Could you expand on why you feel this is putting pressure on everyone to be N-0?

My understanding is that all it takes is a single dependency in your tree to use the default policy `"auto"` whether willingly or not. That would in my mind increase the probability of seeing the warning (or the error if maximum is enabled). This is scenario to me is a pressure to upgrade, and for developers who care about msrv, chasing down (or discontinuing use of dependencies) which accidentally use the default behavior.

> If they are not intending to support that prior versions then they aren't putting in this effort and it exists in an unknown state. To ensure the quality of the experience of users, we want the default case to avoid these unknown states.

What is a specific harm? What does the degraded user experience look like?

I will add that there is quite a bit of:
> Every feature has a cost and we should balance the cost against the value we expect 

While I agree with the sentiment abstractly, to me it doesn't present a particular and plausible upside or downside. That makes it hard to evaluate.

Reply:
I'll also note that I'm not intentionally looking for an infinite philosophical regression of answering the "why". But to me we've still fallen short.

Reply:
> > If they are not intending to support that prior versions then they aren't putting in this effort and it exists in an unknown state. To ensure the quality of the experience of users, we want the default case to avoid these unknown states.
>
> What is a particular harm? What does the degraded user experience look like?

- A user depending on a package without an MSRV and it failing to compile for some strange looking reason. They either abandon the package or go to the package's repo and open an issue, waiting on the maintainer (even better if they search and find an existing issue but that seems rare).  Either the maintainer is experienced with this and can immediately report what is happening (users toolchain is old) or they then have to spend time trying to figure out the situation.
- A user proactively tries to use a package with a compatible MSRV and the build fails for mysterious reasons.  Repeat above.

One of the "super powers" of the cargo / crates.io ecosystem has been the adherence to semver.  For myself, I feel its relatively safe and trivial to upgrade within a compatible release.  This is all to say that we should be expecting the same quality of experience / same bar of quality for all other primary interactions with cargo, including `package.rust-version`.

Reply:
> My understanding is that all it takes is a single dependency in your tree to use the default policy "auto" whether willingly or not. That would in my mind increase the probability of seeing the warning (or the error if maximum is enabled). This is scenario to me is a pressure to upgrade, and for developers who care about msrv, chasing down (or discontinuing use of dependencies) which accidentally use the default behavior.

Could you elaborate on how this is different than today where potential dependencies don't align with you on MSRV policies?

Reply:
Great, are these two items an exhaustive list of upsides you see in favor of adopting `"auto"` by default?

What are the downsides you see?

> Could you elaborate on how this is different than today where potential dependencies don't align with you on MSRV policies?

I'll get back to this once I've picked your brain on the above if you don't mind. I don't want to get into an argument for a particular downside right now, rather I want to know if you think there are any the RFC could've outlined and how they compare to the upsides.

Reply:
> Great, are these two items an exhaustive list of upsides you see in favor of adopting "auto" by default?

Since this has spread out, let me re-summarize and expand.
- Low-maintenance, low churn way for someone to be on "latest", if they choose
- (proposed resolver behavior) Packages with an MSRV will be more likely to get the "right" version 
  - Seeing this warning early highlights a potential mismatch of MSRV policy if using an old version isn't sufficient for you
- (current resolver behavior) Packages building on an MSRV will be more likely to be told what the right version is to use
- Helps percolate out awareness of new toolchain versions for those wanting to stay on closer to latest
- Using unsupported versions requires a more explicit opt-in acknowledging you are in uncharted territories.

Note that this is upsides, not motivations, which can include more items

> What are the downsides you see?

- "Extended MSRV" users will be stuck on older versions than they would have otherwise unless they opt-in to the old behavior (pick new versions somehow and `allow` the incompatible-MSRV diagnostic)

(yes, that is both an upside and a downside)

Reply:
Thanks, is that now what you consider a complete list of upsides for adopting `"auto"`?

So since I think your list of downsides is missing a few items, I'll be more explicit about this question now:

> Could you elaborate on how this is different than today where potential dependencies don't align with you on MSRV policies?

This is pretty much a complete list of unique downsides I see of adopting `"auto"` by default:

Note that all of the following downsides are contingent on the package building correctly on the current compiler version. Whether that is or isn't more prevalent is currently not the issue. If you acknowledge that they at least exist, you can make that argument when refuting them in the RFC:
* Adopting `auto` by default deviates from established cargo behavior when publishing a crate.
* Adopting `auto` by default causes more users to be faced with an error or warning telling them to upgrade Rust in cases where they wouldn't have had to otherwise.
* Adopting `auto` by default will cause more users who cannot easily upgrade to Rust N-0 to have a degraded user experience.
* Adopting `auto` by default will most of the time pick the incorrect rust-version, thereby causing more errors or warnings.

Finally this is yours:
* Adopting `auto` by default will cause users to be stuck on older versions of dependencies unless; for the `rust-version` resolver they opt out from the diagnostic every time they run `cargo update`; or for the `maximum` resolver perform precise updates.

Preferably I'd like to see these explicitly listed, and a cost analysis performed for why we should disregard them over the upsides.

In terms of value statements, there is only one I would like to see added, but I'm not gonna be super stingy about it being omitted:
* The priority of `cargo build` and `cargo update` should be to build or upgrade the package with the current system configuration, not to prompt the user to install a new Rust version.

Basically I believe in the principle that build tools should not be a channel to promote new Rust versions. I think this is in contrast with what you believe since you've listed "Helps percolate out awareness of new toolchain versions for those wanting to stay on closer to latest" as an upside.

If the RFC acknowledges that those downsides exist, and clear explicit written arguments for why the upsides win out that finds consensus I would be pretty much satisfied.

Reply:
> The priority of cargo build and cargo update should be to build or upgrade the package with the current system configuration, not to prompt the user to install a new Rust version.

In my quick glance, I'm not seeing that I talked about what the notice would look like.  1c66807d makes those assumptions more explicit.  This was never intended to shout in people's faces that they need to upgrade but it raises awareness more subtly.

(will get to the rest soon, just wanted to address this one real quick)

Reply:
> Adopting auto by default deviates from established cargo behavior when publishing a crate.

Could you help me understand how you see this changing established behavior?

We aren't changing the default for `package.rust-version` when its unset but changing  `cargo new` to set it (if there isn't a workspace value to inherit).   So only new packages after this change are affected and the user can sees that this is a choice and can remove it.    This is a new behavior that we do a one-time opt-in which I see as different than changing established behavior.

> Adopting auto by default will most of the time pick the incorrect rust-version, thereby causing more errors or warnings.

How is this different than "Adopting auto by default causes more users to be faced with an error or warning telling them to upgrade Rust in cases where they wouldn't have had to otherwise."?

As for what the right rust version is, that comes back to the question of what "support" means and what our expectations are for quality are.  For the latter, that would get into whether the quality of the experience is helped more by using safer (validated) older versions or newer unvalidated (could fail) versions.

Reply:
> Adopting auto by default will cause more users who cannot easily upgrade to Rust N-0 to have a degraded user experience.

Let me see if my re-wording of this works for you.  This would go under the two "Extended MSRV" workflows
- 🟰 Wide-spread use of `package.rust-version = "auto"` (and other proactive MSRV update strategies) reduces the number of package versions available to use with older MSRVs.  However, that pool was tainted with incompatible versions that led to user confusion and maintainer frustration.

(still not sure I've fully captured the nuance and how best this fits within the document)

Reply:
> Could you help me understand how you see this changing established behavior?
> We aren't changing the default for package.rust-version when its unset but changing cargo new to set it (if there isn't a workspace value to inherit). So only new packages after this change are affected and the user can sees that this is a choice and can remove it. This is a new behavior that we do a one-time opt-in which I see as different than changing established behavior.

The established behavior: `cargo new` will not set `rust-version` and `cargo publish` doesn't include `rust-version` if it's not specified.
The proposed behavior: `cargo new` will populate rust-version to `auto` and `cargo publish` will default to `auto` if it's not specified. Every `cargo publish` with this default will publish a crate with `rust-version = $(rustc --version)`.

Is this the proposed new default behavior and if so, is it not a deviation from the established default behavior?

> How is this different than "Adopting auto by default causes more users to be faced with an error or warning telling them to upgrade Rust in cases where they wouldn't have had to otherwise."?

That's true, I started with 6 before culling them. This one slipped through!

Reply:
> The proposed behavior: cargo new will populate rust-version to auto and ~~cargo publish will default to auto if it's not specified.~~

With this proposal, the default when a `Cargo.toml` is missing `package.rust-version` remains the same: there is no defined rust-version / MSRV. 

To make `auto` the default within the manifest schema would be a lot bigger of a deal (to the point of being tied to an Edition) and we would need a way for people to opt out (well, I guess `rust-version = "1.0"` is an opt-out but that carries a different intent).

Changing the default for `Cargo.toml` is only ever talked about as an [alternative](https://github.com/epage/rfcs/blob/msrv/text/3537-msrv-resolver.md#ensuring-the-registry-index-has-rust-version-without-affecting-quality) / maybe something to consider in the future in another RFC.

Reply:
Good to know! While that reduces its impact. Now you can focus on why it should be recorded as the default for new crates. The rationales should overlap and the downsides are the same.